services/venice-controller/src/main/java/com/linkedin/venice/controller/ControllerInstanceTagRefresher.java

```diff
@@ -0,0 +1,82 @@
+package com.linkedin.venice.controller;
+
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.helix.SafeHelixManager;
+import java.util.HashSet;
+import java.util.Set;
+import org.apache.helix.ConfigAccessor;
+import org.apache.helix.PreConnectCallback;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * PreConnectCallback is called after the participant registers its info in the cluster but before marking itself as
+ * LIVEINSTANCE. If the instance is already a participant of the cluster, then we need this logic to update Instance tags
+ * in the case of KSAP assigning this controller participant to another cluster
```

Comment:
nit: Replace mentions of KSAP with generic names. Try to keep the replacements to also not be k8s specific